### PR TITLE
Oldgens: Don't show abilities the Pokemon didn't have at the time

### PR DESCRIPTION
**Pull request by asgdf**
In the teambuilder and in battle tooltips, for example, Pelipper will no longer have Drizzle displayed as a possible ability if it didn't have it in that gen.

By the way, is there a reason we do multi-line conditions by putting the conditional operators at the end of the previous line instead of the start of the next one? I think it's legitimately harder to read.

### CODE REVIEW

**Comment by MacChaeger** [comment]
It might be easier to read if it were indented based on paren depth, but that would likely defeat the purpose of breaking it up in the first place. I definitely agree that it's really hard to parse multiline conditionals that are really complex like the ones that plague the randbats generator.

**Comment by Zarel** [comment]
My solution for multiline conditionals has always been to break it apart. This also lets you name the intermediate value.

```
if (tickets[t].ip === user.latestIp && tickets[t].banned &&
  (!Punishments.sharedIps.has(user.latestIp) ||
    Punishments.sharedIps.has(user.latestIp) && !user.autoconfirmed)) {
```

is bad because it's too much information in one statement, and takes a long time to read. Breaking it up into multiple lines doesn't help very much at all.

```
const checkIp = !(Punishments.sharedIps.has(user.latestIp) && user.autoconfirmed);
if (checkIp && tickets[t].ip === user.latestIp && tickets[t].banned) {
```

names the intermediate value `checkIp`, which makes it clearer what it's for, and also prevents there from being too much code on one line.

**Comment by Zarel** [comment]
You might want a helper function `hasAbility` or something.

**Comment by asgdf** [comment]
Ok so this turned into a bit more of a refactor than planned, but now "which abilities does this pokemon have in this gen" and "does this pokemon have this ability in this gen" run off one central implementation rather than having like 5 implementations all over the place.
This is probably slightly less performant than before (especially because it searches for the correct BattleTeambuilderTable on every call), but I didn't notice any perf issues while testing. Should I just have the caller pass in the table instead of the gen?

**Comment by asgdf** [comment]
Oh wait ``Object.assign()`` is ES6. Do we have an established way of doing that in the client?

**Comment by Zarel** [comment]
`$.extend`

**Comment by Zarel** [comment]
https://api.jquery.com/jQuery.extend/

**Comment by Zarel** [comment]
Also feel free to just polyfill `Object.assign`. I started lightly polyfilling other functions because people kept on breaking the client by using ES6 functions

**Comment by asgdf** [comment]
Alright I just replaced it with $.extend.

File: build-tools/build-indexes [hunk 1]
```diff
@@ -619,6 +619,8 @@ process.stdout.write("Building `data/teambuilder-tables.js`... ");
 		BattleTeambuilderTable[gen].overrideType = overrideType;
 		const overrideAbility = {};
 		BattleTeambuilderTable[gen].overrideAbility = overrideAbility;
+		const removeSecondAbility = {};
```
Comment (Zarel) [inline build-tools/build-indexes]: This seems overkill? I believe `ability.gen == 4 && gen === 3` is the only situation you'd want to remove the second ability?

Comment (asgdf) [reply build-tools/build-indexes]: What motivated me to do this was that from gen 6 to 7, Beartic got Slush Rush, Gigalith got Sand Stream, Pelipper got Drizzle and Vamilluxe got Snow Warning.

However, I just noticed that the gen 3->gen 4 ability gains aren't caught by this, because e.g. ``Dex.mod('gen3').data.Pokedex.hitmontop`` already has both Intimidate and Technician listed despite the latter not existing yet. Is that intentional?

Comment (asgdf) [reply build-tools/build-indexes]: Hm that's annoying, so I will have to add the ability gen check after all. I had hoped to avoid that with this whole ordeal.

**Comment by asgdf** [comment]
Okay, rebased, typescripted these two functions, still seems to work.

**Comment by asgdf** [comment]
Bump - unless you still think the separate table is overkill and I should just hardcode exceptions in hasAbility?

File: build-tools/build-indexes [hunk 2]
```diff
@@ -636,6 +638,11 @@ process.stdout.write("Building `data/teambuilder-tables.js`... ");
 			if (pastEntry.abilities['0'] !== nowEntry.abilities['0']) {
 				overrideAbility[id] = pastEntry.abilities['0'];
 			}
+			// in the gen 3 dex, Pokemon already have the abilities that were added in gen 4
+			const hasAbilityFromNewerGen = Tools.getTemplate(id).gen <= genNum && Tools.getAbility(pastEntry.abilities['1']).gen > genNum;
```
Comment (Zarel) [inline build-tools/build-indexes]: Looks like you chose to add the ability gen in the table builder. Good choice.

File: js/client-teambuilder.js [hunk 3]
```diff
@@ -2795,10 +2795,7 @@
 			} else {
 				set.item = '';
 			}
-			set.ability = template.abilities['0'];
-			var table = BattleTeambuilderTable['gen' + this.curTeam.gen];
-			var overrideAbility = table && table.overrideAbility[template.id];
-			if (overrideAbility) set.ability = overrideAbility;
+			set.ability = Tools.getAbilitiesFor(template.id, this.curTeam.gen)['0'];
```
Comment (Zarel) [inline js/client-teambuilder.js]: Does ability 0 actually change between generations? Ick.

File: src/battle-dex.ts [hunk 4]
```diff
@@ -852,6 +852,38 @@ const Tools = {
 		return type;
 	},
 
+	getAbilitiesFor: function (template: any, gen: number) {
```
Comment (Zarel) [inline src/battle-dex.ts]: `getAbilitiesFor(`, not `getAbilitiesFor: function (`. Look at the surrounding code.

(Ditto `hasAbility`)

File: src/battle-dex.ts [hunk 5]
```diff
@@ -852,6 +852,38 @@ const Tools = {
...
+		template = this.getTemplate(template);
+		if ((!isNaN(gen) && gen < 3) || !template.abilities) return {};
```
Comment (Zarel) [inline src/battle-dex.ts]: `!isNaN`?

I assume the `gen` parameter is intended to be optional? What's the intended behavior when `gen` isn't passed? You probably want to specify a default gen, like:

`getAbilitiesFor: function (template: any, gen = 7) {`

or

`getAbilitiesFor: function (template: any, gen = 0) {`

File: src/battle-dex.ts [hunk 6]
```diff
@@ -852,6 +852,38 @@ const Tools = {
...
+		if (!removeSecondAbility && templAbilities['1']) {
+			abilities['1'] = templAbilities['1'];
+		}
+		if (gen >= 5 && templAbilities['H']) abilities['H'] = templAbilities['H'];
+		if (gen >= 7 && templAbilities['S']) abilities['S'] = templAbilities['S'];
+
+		return abilities;
+	},
+
+	hasAbility: function (template: any, ability: string, gen: number) {
```
Comment (Zarel) [inline src/battle-dex.ts]: If `gen` is optional, it should be specified as `gen?: number`

File: js/client-teambuilder.js (hunk 3 above)
Comment (asgdf) [reply js/client-teambuilder.js]: Gengar being the main case (and Naviathan).

**Comment by asgdf** [comment]
Fixed the issues I think (also replaced all my `var`s with `const`s in the ts file)
